iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,325 @@
+    write: Option<OwnedWriteHalf>,
+    state: State,
+    secret_key: PrivateKey,
+    public_key: PublicKey,
+    cipher: Option<SymmetricEncryptor<E>>,
+    network_addr: Addr<Network<E>>,
+}
+
+impl<E: Encryptor + Send + 'static> Peer<E> {
+    pub fn new(id: PeerId, stream: Option<TcpStream>, state: State, addr: Addr<Network<E>>) -> Self {
+        // P2P encryption primitives
+        let x25519 = X25519Sha256::new();
+        let (public_key, secret_key) = x25519.keypair(None).expect("Unexpected error generating keypair");
+        // If we are connected we take apart stream for two halves.
+        // If we are not connected we save Nones and wait for message to start connecting.
+        match stream {
+            None => {
+                Self { id, read: None, write: None, state, secret_key, public_key, cipher: None, network_addr: addr }
+            }
+            Some(stream) => {
+                let (read, write) = stream.into_split();
+                Self { id, read: Some(read), write: Some(write), state, secret_key, public_key, cipher: None, network_addr: addr }
+            }
+        }
+
+    }
+
+    fn stream(&mut self) -> impl Stream<Item = MessageResult> + Send + 'static {
+        let mut read: OwnedReadHalf = self.read.take().unwrap();
+        stream! {
+            loop {
+                if let Err(e) = readable(&read).await {
+                    yield MessageResult::new_error(e);
+                    break;
+                }
+                let result = match read_message(&mut read).await {
+                    Ok(message) => MessageResult::new_message(message),
+                    Err(e) => MessageResult::new_error(e)
+                };
+
+                yield result;
+            }
+        }
+    }
+
+    async fn handshake(&mut self) -> Result<(), io::Error> {
+        match &self.state {
+            State::Connecting => {
+                // Connecting to peer
+                let addr = self.id.address.clone();
+                let stream = TcpStream::connect(addr.clone()).await;
+                match stream {
+                    Ok(stream) => {
+                        let (read, write) = stream.into_split();
+                        self.read = Some(read);
+                        self.write = Some(write);
+                        self.state = State::ConnectedTo;
+                    }
+                    Err(e) => {
+                        warn!("Could not connect to peer on {}! {}", addr, e);
+                    }
+                }
+            }
+            State::ConnectedTo => {
+                // Send client hello with out public key
+                let mut write_half = self.write.as_mut().unwrap();
+                writable(&write_half).await?;
+                send_client_hello(&mut write_half, self.public_key.0.as_slice()).await?;
+                // Read server hello with node's public key
+                let read_half = self.read.as_mut().unwrap();
```

Comment:
Could you move all match arms to separate functions. They are not connected with each other and represent different stages of connection

Reply:
I mean every match arm into its own function...

Reply:
Done.